### ISSUE TITLE
Too many repetative but not-centralized strings

### ISSUE THREAD

**Issue by tanabi** [fuzzball-muck/fuzzball#490]
As I'm working on the MUF commands, I'm noticing that the error message strings are repeated literally all over the place.

If we were to ever want to do any of the following:

* Change a message
* Localization
* Make these messages dynamic (i.e. load them from a file, make a MUCK command to alter them, etc.)

.... it would be a huge effort to do so. 

I think localization is incredibly unlikely -- as far as I know, there's flat 0 demand for that -- but we should consider at the very least converting the repeticious strings to #defines that live in some logical location or locations -- some strings are truly global like "Permission denied" which surely must occur in almost every file in the source -- while others are pretty local to a .c file.

I don't really have a good plan of action here, yet, but I felt like I should call out the problem and this was beyond a simple TODO message :)  Though I'm pretty sure I've done a TODO or two about this already.

**Comment by wyld-sw** [comment]
Would it be worth our time to make a pre-flight process that automatically checks things like mucker level and argument types before (or right after) each primitive is called? This might inform our eventual interpreter overhaul - and maybe help us find inconsistencies.

I probably would want to consider constructing error messages like:

 "Operand 1: 'thing to match' must be a string."

...based on some sort of configuration. I'd certainly want this to be a fast operation, so it may take more thought.

This could consolidate most CHECKOP and CHECKOFLOW uses, and many beginning checks, I would think.

Silly idea? Too much?

**Comment by tanabi** [comment]
I think going to C++ simplifies this a lot, because the base class of MufPrim or whatever we call it will have some set of standardized checks and you can fire them off -- or perhaps we come up with a scheme that even automates that so that the parameters required are basically "data" that is associated with the primitive (which I think is what you're proposing here).  I don't want to get elaborite at the expense of performance, though, but I don't think there's a lot of value in working too hard on the current C implementation.

However, whether or not intentional, you have given me an idea that we could implement in the C version if we wanted to.  What if we had the MUCK equivalent of "perror"?  Pass in a positional argument and an error code number, emit error.  Then, the strings all reside in fb_perror (or whatever it is called) in (initially) a giant switch statement but perhaps someday we use localization or files.  And then we call things like this:

```
if (oper1.type != string) { // or however you do it, don't remember off hte top of my head
     fb_perror(2, REQUIRE_STRING); // emits: Error: Operand 2 must be a string.
}
```
We could use this function in both C and C++ so I think it makes sense to at least consolidate the error messages, though it doesn't solve redundant checks. (but this ticket is about messages, not checks, so it aligns with the original purpose :D )

**Comment by wyld-sw** [comment]
Yep, that'll work. My first goal was to audit all of the primitives, so we could identify inconsistencies and see where things could be done better. That'll obviously help us whatever language we're in, but it makes sense to wait until we're closer to implementing part of this with objects.

I agree with the fb_perror approach. At one point, not all primitive error messages agreed on if "operand 1" was the highest in the stack, or the lowest. Maybe that's not the case anymore. But if so, we should pick a lane :)

**Comment by tanabi** [comment]
We could remove all ambiguity by making a "calling signature" string part of each prim.  This may be more of a C++ idea, but like...
```
"s i ... i d"
```
Error becomes:

```
String operand expected: [s] i ... i d
```

Note that bracket indicats what string is wrong.  We could even theoretically parse those strings to do arg checking ... but I don't really like the idea of string parsing on EVERY PRIM being run, seems overkillish.  I mean, it probably isn't going to break the bank CPU wise, but it makes me cring a lil anyway.

**Comment by wyld-sw** [comment]
I agree that it's probably C++ (which I'm thinking should be soon after FB7's official release - which I may remember we thought could be relatively soon after documentation is done), but I do like the calling signature idea. Having text like "descriptor to notify" associated might just need be in documentation.

**Comment by tanabi** [comment]
Sure.  Let's crank out these docs then we can do it.  Just have like 10,000 more MUF prims to go :)  I think after the p_*'s a lot of the other files are documented ... I did the first files in a weird order (like I think wiz.c is done, for instance).

**Comment by PointsDragon** [comment]
While I agree that i18n is unlikely, there is a lot of call for unicode, so treating all strings as data is a good idea, and consolidating ALL string constants to a strings file just makes a LOT of other work easier down the line.
